### PR TITLE
Upgrade to Apache Tuweni 2.0

### PR DESCRIPTION
**Pull request by atoulme**
Signed-off-by: Antoine Toulme <antoine@lunar-ocean.com>

## PR description
This is a test PR against Apache Tuweni's latest release candidate version 2.

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -105,6 +105,7 @@ allprojects {
   repositories {
     jcenter()
     mavenCentral()
+    maven { url "https://repository.apache.org/content/repositories/orgapachetuweni-1020" }
```
Comment (shemnon) [inline build.gradle:108]: Is this just needed until the 2.0.0 release is approved?

Comment (atoulme) [reply build.gradle:108]: Indeed. This is the staged repository for the current RC.

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,7 +4,7 @@
 
 ### Additions and Improvements
 - `priv_call` now uses NO_TRACING OperationTracer implementation which improves memory usage [\#2482](https://github.com/hyperledger/besu/pull/2482)
-
+- Migration to Apache Tuweni 2.0 [\#2376](https://github.com/hyperledger/besu/pull/2376)
```
Comment (shemnon) [inline CHANGELOG.md:7]: Needs newline or markdown won't render correctly.

**Comment by shemnon** [comment]
While a lot of `toBytes` were removed, there are a few calls to `UInt256.fromBytes` added in some critical spots.  Do you have any measures on performance? i.e. will the `toBytes` removal gains dominate the new `fromBytes` calls?

**Comment by atoulme** [comment]
It does seem reference tests apply just as fast if not faster but that’s hardly a fair approach. I can look at running a goerli full sync and see if it is faster. I’ll look for the call sites you mention, maybe there’s obvious gains too.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/vm/Words.java [hunk 3]
```diff
@@ -31,10 +32,10 @@ private Words() {}
    * @return A VM word containing {@code address} (left-padded as according to the VM specification
    *     (Appendix H. of the Yellow paper)).
    */
-  public static Bytes32 fromAddress(final Address address) {
+  public static UInt256 fromAddress(final Address address) {
```
Comment (garyschulte) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/vm/Words.java:35]: seems like toAddress should take UInt256 for consistency sake.  non-blocking

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/vm/operations/GtOperation.java [hunk 4]
```diff
@@ -29,11 +28,10 @@ public GtOperation(final GasCalculator gasCalculator) {
...
   @Override
   public OperationResult executeFixedCostOperation(final MessageFrame frame, final EVM evm) {
-    final UInt256 value0 = UInt256.fromBytes(frame.popStackItem());
-    final UInt256 value1 = UInt256.fromBytes(frame.popStackItem());
+    final UInt256 value0 = frame.popStackItem();
+    final UInt256 value1 = frame.popStackItem();
 
-    final Bytes32 result =
-        (value0.compareTo(value1) > 0 ? UInt256.valueOf(1) : UInt256.ZERO).toBytes();
+    final UInt256 result = (value0.compareTo(value1) > 0 ? UInt256.valueOf(1) : UInt256.ZERO);
```
Comment (garyschulte) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/vm/operations/GtOperation.java:34]: UInt256.ONE is static, might as well clean up the place while we are here :)

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/vm/operations/SDivOperation.java [hunk 5]
```diff
@@ -32,12 +32,12 @@ public SDivOperation(final GasCalculator gasCalculator) {
...
-    final Bytes32 value0 = frame.popStackItem();
-    final Bytes32 value1 = frame.popStackItem();
+    final UInt256 value0 = frame.popStackItem();
+    final UInt256 value1 = frame.popStackItem();
     if (value1.isZero()) {
-      frame.pushStackItem(Bytes32.ZERO);
+      frame.pushStackItem(UInt256.ZERO);
     } else {
-      final BigInteger result = value0.toBigInteger().divide(value1.toBigInteger());
+      final BigInteger result = value0.toSignedBigInteger().divide(value1.toSignedBigInteger());
```
Comment (garyschulte) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/vm/operations/SDivOperation.java:40]: just out of curiosity, if value0 and value1 are UInt256, why `toSignedBigInteger` for the division?

**Comment by shemnon** [comment]
Re: benchmarks 
I think a reference test run before and after on the same hardware would give the numbers I am looking for.  Quicker than a goerli full sync.  And I don't expect it would diverge from the reference tests.

**Comment by shemnon** [comment]
I also would _not_ trust CI builds for the benchmark numbers.

**Comment by garyschulte** [comment]
> Re: benchmarks
> I think a reference test run before and after on the same hardware would give the numbers I am looking for. Quicker than a goerli full sync. And I don't expect it would diverge from the reference tests.

running general state tests locally against this PR and 21.7.0, it is pretty close with a slight edge to tuweni-2:
```
   this pr:
      sh ./dretesteth.sh -t GeneralStateTests ... 0.25s user 0.16s system 0% cpu 12:05.32 total
      sh ./dretesteth.sh -t GeneralStateTests ... 0.24s user 0.15s system 0% cpu 11:19.69 total
      sh ./dretesteth.sh -t GeneralStateTests ... 0.25s user 0.15s system 0% cpu 11:55.20 total

   21.7.0
      sh ./dretesteth.sh -t GeneralStateTests ... 0.27s user 0.23s system 0% cpu 13:36.72 total
      sh ./dretesteth.sh -t GeneralStateTests ... 0.25s user 0.17s system 0% cpu 11:28.28 total
      sh ./dretesteth.sh -t GeneralStateTests ... 0.25s user 0.17s system 0% cpu 11:27.22 total
  ```
  
  the 13:36 outlier had a zoom call running near the end of it, but the others were run with the machine unencumbered.

**Comment by atoulme** [comment]
Thanks! I have to look at memory usage as well.

**Comment by shemnon** [comment]
Sight improvement is fine.  My concern was that this is the kind of a code location where a regression could have a major impact.  Evidence indicates this is not the case.

**Review by shemnon — APPROVED** [review]

**Comment by atoulme** [comment]
Ok will address reviews. Thanks folks

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/vm/Words.java (hunk 3 above)
Comment (atoulme) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/vm/Words.java:35]: UInt256 now implements Bytes32, so it's ok that way around.

**Comment by atoulme** [comment]
Just waiting for CHANGELOG updates on master, and then I can resolve conflicts and get this in.

**Pull request merged**